packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticMinter.sol

```diff
@@ -0,0 +1,87 @@
+
+/**
+ * @notice An example use case of the OptimisticRewarder in use by a contract that allows users to stake an ERC20 to
+ * earn rewards.
+ */
+contract OptimisticMintr is Lockable {
+    using SafeERC20 for IERC20;
+
+    // Optimistic rewarder contract used to pay out user rewards.
+    OptimisticRewarderBase public optimisticRewarder;
+
+    // Staked ERC20 token.
+    IERC20 public stakedToken;
+
+    // Balances by tokenId.
+    mapping(uint256 => uint256) public balances;
+
+    /**
+     * @notice Constructor.
+     * @param _optimisticRewarder Optimistic rewarder contract used to pay out user rewards.
+     * @param _stakedToken staked ERC20 token.
+     */
+    constructor(OptimisticRewarderBase _optimisticRewarder, IERC20 _stakedToken) {
+        optimisticRewarder = _optimisticRewarder;
+        stakedToken = _stakedToken;
+    }
+
+    modifier onlyTokenOwner(uint256 tokenId) {
+        require(optimisticRewarder.ownerOf(tokenId) == msg.sender, "caller != token owner");
+        _;
+    }
+
+    /**
+     * @notice Deposit the staked token into the contract and mint a fresh token to manage the position.
+     * @param tokenId the tokenId that will own this liquidity. User must be the owner of this tokenId.
+     * @param amount the amount of the ERC20 to deposit.
+     * @return tokenId the token id for the freshly minted token.
+     */
+    function depositNew(uint256 amount) public nonReentrant returns (uint256 tokenId) {
+        tokenId = optimisticRewarder.mint(msg.sender, msg.data);
+        _depositFor(tokenId, amount);
+    }
+
+    /**
+     * @notice Deposit the staked token into the contract.
+     * @param tokenId the tokenId that will own this liquidity. User must be the owner of this tokenId.
+     * @param amount the amount of the ERC20 to deposit.
+     */
+    function deposit(uint256 tokenId, uint256 amount) public nonReentrant onlyTokenOwner(tokenId) {
+        _depositFor(tokenId, amount);
+    }
+
+    /**
+     * @notice Deposit staked tokens on behalf of a tokenId that the user may not control.
+     * @param tokenId the tokenId that will own this liquidity.
+     * @param amount the amount of the ERC20 to deposit.
+     */
+    function depositFor(uint256 tokenId, uint256 amount) public nonReentrant {
+        _depositFor(tokenId, amount);
+    }
+
+    /**
+     * @notice Withdraw the staked tokens.
+     * @param tokenId the tokenId that owns this liquidity. User must own this tokenId.
+     * @param amount the amount of the ERC20 to withdraw.
+     */
+    function withdraw(uint256 tokenId, uint256 amount) public nonReentrant onlyTokenOwner(tokenId) {
+        balances[tokenId] -= amount;
+        stakedToken.safeTransfer(msg.sender, amount);
+        optimisticRewarder.updateToken(tokenId, msg.data);
+    }
+
+    function _depositFor(uint256 tokenId, uint256 amount) public {
+        balances[tokenId] += amount;
+        stakedToken.safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
nit: you should do the transfer before incrementing the local variable.

Reply:
Yeah, I was just following the Check-Effects-Interaction pattern.

Reply:
@nick you right. I was confused for a hot second there. you should always leave external calls for as late as posible inside a function.

However, this does _not_ have reentrency guards on it. if the intention was for this to only be called in the depositNew and deposit functions then make this internal.

Reply:
@chrismaree that's a great catch. This function _is_ supposed to be internal, that's a bug. Thanks for the call out!!

Reply:
Fixed!